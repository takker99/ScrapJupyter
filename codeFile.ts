import { Scrapbox } from "./deps/scrapbox.ts";
import { encodeTitle } from "./utils.ts";
declare const scrapbox: Scrapbox;

type File = {
  filename?: string;
  dir: string;
  lang: string;
  /** コードブロックの開始行のid */ startIds: string[];
  lines: string[];
};

export function getCodeFiles() {
  const codeBlocks =
    scrapbox.Page.lines?.flatMap((line) => "codeBlock" in line ? [line] : []) ??
      [];
  return codeBlocks.reduce((acc: File[], { codeBlock, text, id }) => {
    const sameFileIndex = acc.findIndex(({ filename }) =>
      filename !== undefined && filename === codeBlock.filename
    );
    // code blockの先頭かつ新しいコードブロックのときのみ新しいfileを追加する
    if (codeBlock.start && sameFileIndex < 0) {
      return [...acc, {
        filename: codeBlock.filename,
        dir: `https://scrpabox.io/api/code/${scrapbox.Project.name}/${
          encodeTitle(scrapbox.Page.title ?? "")
        }`,
        lang: codeBlock.lang,
        startIds: [id],
        lines: [] as string[],
      }];
    }

    if (codeBlock.start) {
      acc.at(sameFileIndex)?.startIds?.push?.(id);
    } else {
      // 既存のコードブロックもしくは末尾のコードブロックに追記する
      acc.at(sameFileIndex)?.lines?.push?.(text);
    }

    return acc;
  }, []);
}
