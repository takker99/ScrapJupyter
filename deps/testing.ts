export * from "jsr:@std/assert@1/equals";
export * from "jsr:@std/testing@1/snapshot";
