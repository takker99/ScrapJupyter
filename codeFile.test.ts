import { getCodeFiles } from "./codeFile.ts";
import { assertSnapshot } from "./deps/testing.ts";
import json from "./sample-page.json" assert { type: "json" };

Deno.test("getCodeFiles()", async (t) => {
  await assertSnapshot(t, getCodeFiles("villagepump", json.title, json.lines));
});
