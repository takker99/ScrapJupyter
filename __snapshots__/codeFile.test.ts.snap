export const snapshot = {};

snapshot[`getCodeFiles() 1`] = `
[Map Iterator] {
  {
    lang: "css",
    path: "https://scrapbox.io/api/code/villagepump/%E3%83%AA%E3%83%B3%E3%82%AF%E3%82%92%E5%87%BA%E5%85%B8%E3%82%A2%E3%82%A4%E3%82%B3%E3%83%B3%E3%81%AB%E3%81%99%E3%82%8BUserCSS/style.css",
    startIds: [
      "61596b221280f000007378f1",
      "6157ea821280f00000ad879f",
      "61580c691280f0000097e632",
      "61fe4ade1280f00000a67ccc",
      "61580c881280f0000097e634",
      "61fe3ffd1280f000008faa49",
    ],
  },
  {
    lang: "css",
    path: "https://scrapbox.io/api/code/villagepump/%E3%83%AA%E3%83%B3%E3%82%AF%E3%82%92%E5%87%BA%E5%85%B8%E3%82%A2%E3%82%A4%E3%82%B3%E3%83%B3%E3%81%AB%E3%81%99%E3%82%8BUserCSS/style.css.disabled",
    startIds: [
      "61596b3d1280f0000073793a",
    ],
  },
  {
    lang: "diff",
    path: "https://scrapbox.io/api/code/villagepump/%E3%83%AA%E3%83%B3%E3%82%AF%E3%82%92%E5%87%BA%E5%85%B8%E3%82%A2%E3%82%A4%E3%82%B3%E3%83%B3%E3%81%AB%E3%81%99%E3%82%8BUserCSS/%E5%B7%AE%E5%88%86.diff",
    startIds: [
      "61fe4a9b1280f00000a67cbd",
    ],
  },
  {
    lang: "diff",
    path: "https://scrapbox.io/api/code/villagepump/%E3%83%AA%E3%83%B3%E3%82%AF%E3%82%92%E5%87%BA%E5%85%B8%E3%82%A2%E3%82%A4%E3%82%B3%E3%83%B3%E3%81%AB%E3%81%99%E3%82%8BUserCSS/%E5%B7%AE%E5%88%862.diff",
    startIds: [
      "61fe4af31280f00000a67cd2",
    ],
  },
}
`;
